File: app/src/main/java/org/surfrider/surfnet/detection/tflite/YoloDetector.java
/* Copyright 2019 The TensorFlow Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/

package org.surfrider.surfnet.detection.tflite;

import android.content.res.AssetManager;
import android.graphics.Bitmap;
import android.graphics.RectF;

import org.tensorflow.lite.Interpreter;
import org.tensorflow.lite.Tensor;
import org.surfrider.surfnet.detection.env.Utils;
import org.tensorflow.lite.gpu.GpuDelegate;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.MappedByteBuffer;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;
import java.util.PriorityQueue;
import java.util.Vector;

import timber.log.Timber;


/**
 * Wrapper for frozen detection models trained using the Tensorflow Object Detection API:
 * - <a href="https://github.com/tensorflow/models/tree/master/research/object_detection">https://github.com/tensorflow/models/tree/master/research/object_detection</a>
 * where you can find the training code.
 * <p>
 * To use pretrained models in the API or convert to TF Lite models, please see docs for details:
 * - <a href="https://github.com/tensorflow/models/blob/master/research/object_detection/g3doc/detection_model_zoo.md">https://github.com/tensorflow/models/blob/master/research/object_detection/g3doc/detection_model_zoo.md</a>
 * - <a href="https://github.com/tensorflow/models/blob/master/research/object_detection/g3doc/running_on_mobile_tensorflowlite.md#running-our-model-on-android">https://github.com/tensorflow/models/blob/master/research/object_detection/g3doc/running_on_mobile_tensorflowlite.md#running-our-model-on-android</a>
 */
public class YoloDetector implements Detector {

    /**
     * Initializes a native TensorFlow session for classifying images.
     *
     * @param assetManager  The asset manager to be used to load assets.
     * @param modelFilename The filepath of the model GraphDef protocol buffer.
     * @param labelFilename The filepath of label file for classes.
     * @param isQuantized   Boolean representing model is quantized or not
     */
    public static YoloDetector create(
            final AssetManager assetManager,
            final String modelFilename,
            final String labelFilename,
            final float confThreshold,
            final boolean isQuantized,
            final boolean isV8,
            final int inputSize)
            throws IOException {
        final YoloDetector d = new YoloDetector();

        String actualFilename = labelFilename.split("file:///android_asset/")[1];
        InputStream labelsInput = assetManager.open(actualFilename);
        BufferedReader br = new BufferedReader(new InputStreamReader(labelsInput));
        String line;
        while ((line = br.readLine()) != null) {
            Timber.w(line);
            d.labels.add(line);
        }
        br.close();

        try {
            Interpreter.Options options = (new Interpreter.Options());

            options.setNumThreads(NUM_THREADS);
            d.tfliteModel = Utils.loadModelFile(assetManager, modelFilename);
            d.tfLite = new Interpreter(d.tfliteModel, options);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }

        d.confThreshold = confThreshold;
        d.isModelQuantized = isQuantized;
        d.isV8 = isV8;
        // Pre-allocate buffers.
        int numBytesPerChannel;
        if (isQuantized) {
            numBytesPerChannel = 1; // Quantized
        } else {
            numBytesPerChannel = 4; // Floating point
        }
        d.INPUT_SIZE = inputSize;
        d.imgData = ByteBuffer.allocateDirect(d.INPUT_SIZE * d.INPUT_SIZE * 3 * numBytesPerChannel);
        d.imgData.order(ByteOrder.nativeOrder());

        double calculation = Math.pow((inputSize / 32d), 2) + Math.pow((inputSize / 16d), 2) + Math.pow((inputSize / 8d), 2);
        if(!isV8) {
            // yolov5 case (20² + 40² + 80²)*3 = 25200
            d.output_box = (int) (calculation * 3);
        } else {
            // yolov8 case (20² + 40² + 80²) = 8400
            d.output_box = (int) calculation;
        }
        if (d.isModelQuantized){
            Tensor oupten = d.tfLite.getOutputTensor(0);
            d.oup_scale = oupten.quantizationParams().getScale();
            d.oup_zero_point = oupten.quantizationParams().getZeroPoint();
        }

        int[] shape = d.tfLite.getOutputTensor(0).shape();
        Timber.i("out shape ==== %s", Arrays.toString(shape));
        int numClass;
        if(!isV8) {
            // yolov5 case: (1, num_anchors, num_class+5)
            numClass = shape[shape.length - 1] - 5;
            d.outData = ByteBuffer.allocateDirect(d.output_box * (numClass + 5) * numBytesPerChannel);
        } else {
            // yolov8 case: (1, num_class+4, num_anchors)
            numClass = shape[shape.length - 2] - 4;
            d.outData = ByteBuffer.allocateDirect(d.output_box * (numClass + 4) * numBytesPerChannel);
        }
        d.numClass = numClass;

        d.outData.order(ByteOrder.nativeOrder());
        return d;
    }

    public int getInputSize() {
        return INPUT_SIZE;
    }

    public void setNumThreads(int num_threads) {
    }

    private void recreateInterpreter() {
        if (tfLite != null) {
            tfLite.close();
            tfLite = new Interpreter(tfliteModel, tfliteOptions);
        }
    }

    public void useGpu() {
        if (gpuDelegate == null) {
            gpuDelegate = new GpuDelegate();
            tfliteOptions.addDelegate(gpuDelegate);
            recreateInterpreter();
        }
    }

    public void useCPU() {
        recreateInterpreter();
    }

    //config yolo
    private int INPUT_SIZE = -1;

    private  int output_box;

    // Number of threads in the java app
    private static final int NUM_THREADS = 1;

    private boolean isModelQuantized;

    private boolean isV8;

    private float confThreshold;

    /** holds a gpu delegate */
    GpuDelegate gpuDelegate = null;

    /** The loaded TensorFlow Lite model. */
    private MappedByteBuffer tfliteModel;

    /** Options for configuring the Interpreter. */
    private final Interpreter.Options tfliteOptions = new Interpreter.Options();

    // Config values.

    // Pre-allocated buffers.
    private final Vector<String> labels = new Vector<>();
    private ByteBuffer imgData;
    private ByteBuffer outData;

    private Interpreter tfLite;
    private float oup_scale;
    private int oup_zero_point;
    private int numClass;
    private YoloDetector() {
    }

    //non maximum suppression
    protected ArrayList<Recognition> nms(ArrayList<Recognition> list) {
        ArrayList<Recognition> nmsList = new ArrayList<>();

        for (int k = 0; k < labels.size(); k++) {
            //1.find max confidence per class
            PriorityQueue<Recognition> pq =
                    new PriorityQueue<>(
                            50,
                            (lhs, rhs) -> {
                                // Intentionally reversed to put high confidence at the head of the queue.
                                return Float.compare(rhs.getConfidence(), lhs.getConfidence());
                            });

            for (int i = 0; i < list.size(); ++i) {
                if (list.get(i).getDetectedClass() == k) {
                    pq.add(list.get(i));
                }
            }

            //2.do non maximum suppression
            while (pq.size() > 0) {
                //insert detection with max confidence
                Recognition[] a = new Recognition[pq.size()];
                Recognition[] detections = pq.toArray(a);
                Recognition max = detections[0];
                nmsList.add(max);
                pq.clear();

                for (int j = 1; j < detections.length; j++) {
                    Recognition detection = detections[j];
                    RectF b = detection.getLocation();
                    if (box_iou(max.getLocation(), b) < mNmsThresh) {
                        pq.add(detection);
                    }
                }
            }
        }
        return nmsList;
    }

    protected float mNmsThresh = 0.6f;

    protected float box_iou(RectF a, RectF b) {
        return box_intersection(a, b) / box_union(a, b);
    }

    protected float box_intersection(RectF a, RectF b) {
        float w = overlap((a.left + a.right) / 2, a.right - a.left,
                (b.left + b.right) / 2, b.right - b.left);
        float h = overlap((a.top + a.bottom) / 2, a.bottom - a.top,
                (b.top + b.bottom) / 2, b.bottom - b.top);
        if (w < 0 || h < 0) return 0;
        return w * h;
    }

    protected float box_union(RectF a, RectF b) {
        float i = box_intersection(a, b);
        return (a.right - a.left) * (a.bottom - a.top) + (b.right - b.left) * (b.bottom - b.top) - i;
    }

    protected float overlap(float x1, float w1, float x2, float w2) {
        float l1 = x1 - w1 / 2;
        float l2 = x2 - w2 / 2;
        float left = Math.max(l1, l2);
        float r1 = x1 + w1 / 2;
        float r2 = x2 + w2 / 2;
        float right = Math.min(r1, r2);
        return right - left;
    }
    public ArrayList<Recognition> recognizeImage(Bitmap bitmap) {
        Map<Integer, Object> outputMap = new HashMap<>();

        outData.rewind();
        outputMap.put(0, outData);
        // Timber.tag("YoloDetector").d("mObjThresh: %s", getObjThresh());

        Object[] inputArray = {imgData};
        tfLite.runForMultipleInputsOutputs(inputArray, outputMap);

        ByteBuffer byteBuffer = (ByteBuffer) outputMap.get(0);
        assert byteBuffer != null;
        byteBuffer.rewind();

        ArrayList<Recognition> detections = new ArrayList<>();

        float[][][] out = new float[1][output_box][numClass + 5];
        if(!isV8) {
            // Timber.tag("YoloDetector").d("out[0] detect start");
            for (int i = 0; i < output_box; ++i) {
                for (int j = 0; j < numClass + 5; ++j) {
                    if (isModelQuantized) {
                        out[0][i][j] = oup_scale * (((int) byteBuffer.get() & 0xFF) - oup_zero_point);
                    } else {
                        out[0][i][j] = byteBuffer.getFloat();
                    }
                }
                // Denormalize xywh
                for (int j = 0; j < 4; ++j) {
                    out[0][i][j] *= getInputSize();
                }
            }
        } else {
            // switch the way we span through the bytebuffer
            // Timber.tag("YoloDetector V8").d("out[0] detect start");
            for (int j = 0; j < numClass + 4; ++j) {
                for (int i = 0; i < output_box; ++i) {
                    if (isModelQuantized) {
                        out[0][i][j] = oup_scale * (((int) byteBuffer.get() & 0xFF) - oup_zero_point);
                    } else {
                        out[0][i][j] = byteBuffer.getFloat();
                    }
                }
            }
            // no need to denormalize for yolov8
        }
        for (int i = 0; i < output_box; ++i) {
            final int offset = 0;
            float confidence = 1.0f;
            if(!isV8) {
                // confidence only valid for yolov5
                confidence = out[0][i][4];
            }
            int detectedClass = -1;
            float maxClass = 0;

            final float[] classes = new float[labels.size()];
            for (int c = 0; c < labels.size(); ++c) {
                if(!isV8) {
                    classes[c] = out[0][i][5 + c];
                } else {
                    classes[c] = out[0][i][4 + c];
                }
            }

            for (int c = 0; c < labels.size(); ++c) {
                if (classes[c] > maxClass) {
                    detectedClass = c;
                    maxClass = classes[c];
                }
            }

            final float confidenceInClass = maxClass * confidence;
            if (confidenceInClass > confThreshold) {
                final float xPos = out[0][i][0];
                final float yPos = out[0][i][1];

                final float w = out[0][i][2];
                final float h = out[0][i][3];
                // Timber.tag("YoloDetector").d(Float.toString(xPos) + ',' + yPos + ',' + w + ',' + h);

                final RectF rect =
                        new RectF(
                                Math.max(0, xPos - w / 2),
                                Math.max(0, yPos - h / 2),
                                Math.min(bitmap.getWidth() - 1, xPos + w / 2),
                                Math.min(bitmap.getHeight() - 1, yPos + h / 2));
                detections.add(new Recognition("" + offset, labels.get(detectedClass),
                        confidenceInClass, rect, detectedClass));
            }
        }
        // Timber.tag("YoloDetector").d("detect end");

        return nms(detections);
    }
}

File: app/src/main/java/org/surfrider/surfnet/detection/tflite/YoloDetector.kt
/* Copyright 2019 The TensorFlow Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/
package org.surfrider.surfnet.detection.tflite

import android.content.Context
import android.content.res.AssetManager
import android.graphics.Bitmap
import android.graphics.RectF
import android.os.SystemClock
import org.opencv.core.CvType
import org.opencv.core.Mat
import org.opencv.core.Rect
import org.opencv.imgproc.Imgproc
import org.surfrider.surfnet.detection.env.Utils.loadModelFile
import org.surfrider.surfnet.detection.tflite.Detector.Recognition
import org.tensorflow.lite.Interpreter
import org.tensorflow.lite.gpu.CompatibilityList
import org.tensorflow.lite.gpu.GpuDelegate
import org.tensorflow.lite.gpu.GpuDelegateFactory.Options.INFERENCE_PREFERENCE_FAST_SINGLE_ANSWER
import org.tensorflow.lite.gpu.GpuDelegateFactory.Options.INFERENCE_PREFERENCE_SUSTAINED_SPEED
import timber.log.Timber
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.FloatBuffer
import java.nio.MappedByteBuffer
import java.util.Vector
import kotlin.math.max
import kotlin.math.pow

class YoloDetector private constructor() : Detector {
    override fun setNumThreads(numThreads: Int) {}

    // Float model
    private val IMAGE_MEAN = 0f
    private val IMAGE_STD = 255.0f

    //config yolo
    var inputSize = -1
    var numMasks = 0
    var resolutionMaskW = 0
    var resolutionMaskH = 0
    private var outputBox = 0
    private var isModelQuantized = false
    private var modelType = ""
    private var confThreshold = 0f
    private var scalingFactor = 1.0f

    /** The loaded TensorFlow Lite model.  */
    private var tfliteModel: MappedByteBuffer? = null

    // Config values.
    private val labels = Vector<String>()
    private lateinit var intValues: IntArray
    private lateinit var imgFloatArray: FloatArray
    private lateinit var imgData: ByteBuffer
    private lateinit var outData: ByteBuffer
    private lateinit var outData2: ByteBuffer
    private lateinit var masks: Mat
    private lateinit var predsCV: Mat
    private lateinit var preds: Array<FloatArray>
    private var tfLite: Interpreter? = null
    private var inp_scale = 0f
    private var inp_zero_point = 0
    private var oup_scale = 0f
    private var oup_zero_point = 0
    private var numClass = 0
    private lateinit var context: Context

    private fun convertMatToByteBuffer(mat: Mat) {
        imgData.rewind()
        Imgproc.cvtColor(mat, mat, Imgproc.COLOR_RGBA2RGB)
        val floatMat = Mat()
        mat.convertTo(floatMat, CvType.CV_32FC3, 1.0 / IMAGE_STD, 1.0 * IMAGE_MEAN)
        val newMat = floatMat.reshape(1, 1)
        newMat.get(0, 0, imgFloatArray)
        imgData.asFloatBuffer().put(imgFloatArray)
    }

    private fun processTFoutput(out: Mat, width:Int, height:Int): ArrayList<Recognition> {
        // Assumes a matrix of shape (output_box, 4 + numClass) or (output_box, 4 + numClass + numMask)
        val detections = ArrayList<Recognition>()
        val classes = FloatArray(numClass)
        val boxes = FloatArray(4)
        for (i in 0 until outputBox) {
            out.get(i, 4, classes)
            var detectedClass = -1
            var maxClass = 0.0f
            for (c in 0 until numClass) {
                if (classes[c] > maxClass) {
                    detectedClass = c
                    maxClass = classes[c]
                    if (maxClass > 0.5) {
                        break
                    }
                }
            }
            val confidenceInClass = maxClass
            if (confidenceInClass > confThreshold) {
                out.get(i, 0, boxes)
                val xPos = boxes[0] * scalingFactor
                val yPos = boxes[1] * scalingFactor
                val w = boxes[2] * scalingFactor
                val h = boxes[3] * scalingFactor
                val rect = RectF(
                    max(0.0f, xPos - w / 2),
                    max(0.0f, yPos - h / 2),
                    kotlin.math.min(width - 1.0f, xPos + w / 2),
                    kotlin.math.min(height - 1.0f, yPos + h / 2)
                )
                detections.add(
                    Recognition(labels[detectedClass],
                        confidenceInClass, rect, i,null, detectedClass, null
                    )
                )
            }
        }
        return detections
    }

    override fun recognizeImage(frame: Mat): ArrayList<Recognition?>? {
        return if (modelType == "segmentation") {
            segmentImage(frame)
        } else {
            detectImage(frame)
        }
    }
    private fun detectImage(frame: Mat): ArrayList<Recognition?>? {
        val preprocessTime = SystemClock.uptimeMillis()
        convertMatToByteBuffer(frame)
        val outputMap: MutableMap<Int, Any?> =
            HashMap()
        outData.rewind()
        outputMap[0] = outData
        val inputArray = arrayOf<Any?>(imgData)
        val startTime = SystemClock.uptimeMillis()
        tfLite!!.runForMultipleInputsOutputs(inputArray, outputMap)
        val tfTime = SystemClock.uptimeMillis()

        val byteBuffer = outputMap[0] as ByteBuffer?
        byteBuffer!!.rewind()

        // switch the way we span through the bytebuffer
        if(isModelQuantized) {
            for (j in 0 until 4 + numClass) {
                for (i in 0 until outputBox) {
                    preds[i][j] =
                        oup_scale * ((byteBuffer.get().toInt() and 0xFF) - oup_zero_point)
                }
            }
        } else {
            predsCV = Mat(4+numClass, outputBox, CvType.CV_32F, byteBuffer).t()
        }
        val allocateTime = SystemClock.uptimeMillis()
        val detections = processTFoutput(predsCV, frame.rows(), frame.cols())
        val processOutputTime = SystemClock.uptimeMillis()
        if(DEBUG_PROFILING)
            Timber.i("pre: ${startTime - preprocessTime}" +
                " tf: ${tfTime - startTime}" +
                " allocate: ${allocateTime - tfTime}" +
                " outProcess: ${processOutputTime - allocateTime}")
        return DetectorUtils.nms(detections, numClass)
    }

    private fun segmentImage(frame: Mat): ArrayList<Recognition?>? {
        val preprocessTime = SystemClock.uptimeMillis()
        convertMatToByteBuffer(frame)
        val outputMap: MutableMap<Int, Any?> =
            HashMap()
        outData.rewind()
        outData2.rewind()
        outputMap[0] = outData
        outputMap[1] = outData2
        val inputArray = arrayOf<Any?>(imgData)
        val startTime = SystemClock.uptimeMillis()
        tfLite!!.runForMultipleInputsOutputs(inputArray, outputMap)
        val tfTime = SystemClock.uptimeMillis()

        val byteBuffer = outputMap[0] as ByteBuffer?
        val byteBuffer2 = outputMap[1] as ByteBuffer?
        byteBuffer!!.rewind()
        byteBuffer2!!.rewind()


        // Get first output
        if(isModelQuantized) {
            for (j in 0 until 4 + numClass + numMasks) {
                for (i in 0 until outputBox) {
                    preds[i][j] =
                        oup_scale * ((byteBuffer.get().toInt() and 0xFF) - oup_zero_point)
                }
            }
        } else {
            predsCV = Mat(4+numClass+numMasks, outputBox, CvType.CV_32F, byteBuffer).t()
        }

        // Get second output, careful as it shape is HWC
        val allocateTime = SystemClock.uptimeMillis()
        masks = Mat(resolutionMaskH* resolutionMaskW, numMasks, CvType.CV_32F, byteBuffer2).t()
        val allocateMaskTime = SystemClock.uptimeMillis()
        val detections = processTFoutput(predsCV, frame.rows(), frame.cols())
        val processOutputTime = SystemClock.uptimeMillis()
        val indicesToKeep = DetectorUtils.nmsIndices(detections, numClass)
        val nmsTime = SystemClock.uptimeMillis()

        val newDetections = indicesToKeep
            .map { computeMask(detections[it], predsCV.submat(detections[it].maskIdx, detections[it].maskIdx + 1, 4+numClass, 4+numClass+numMasks))}
//        preds[detections[it].maskIdx].sliceArray(4+numClass..<4+numClass+numMasks)) }

            .toCollection(ArrayList())
        val postprocessTime = SystemClock.uptimeMillis()
        if(DEBUG_PROFILING)
            Timber.i("pre: ${startTime - preprocessTime}" +
                    " tf: ${tfTime - startTime}" +
                    " allocate: ${allocateTime - tfTime}" +
                    " allocate mask: ${allocateMaskTime - allocateTime}" +
                    " outProcess: ${processOutputTime - allocateMaskTime}" +
                    " nms: ${nmsTime - processOutputTime}" +
                    " mask: ${postprocessTime - nmsTime}")
        return newDetections
    }

    private fun computeMask(det: Recognition, maskWeights: Mat): Recognition? {
        // Performs inplace update of Mask
        val rect = Rect(det.location.left.toInt() / MASK_SCALE_FACTOR,
            det.location.top.toInt() / MASK_SCALE_FACTOR,
            det.location.width().toInt() / MASK_SCALE_FACTOR,
            det.location.height().toInt() / MASK_SCALE_FACTOR)

        det.mask = DetectorUtils.weightedSumOfMasks(masks, maskWeights, resolutionMaskH, rect)
        return det as Recognition?
    }

    companion object {
        @Throws(IOException::class)
        fun create(
            assetManager: AssetManager,
            modelFilename: String?,
            labelFilename: String,
            confThreshold: Float,
            isQuantized: Boolean,
            modelType: String,
            inputSize: Int,
            ctx: Context
        ): YoloDetector {
            val d = YoloDetector()
            d.context = ctx
            val actualFilename = labelFilename.split("file:///android_asset/".toRegex())
                .dropLastWhile { it.isEmpty() }
                .toTypedArray()[1]
            Timber.i("actual File name $actualFilename")
            val labelsInput = assetManager.open(actualFilename)
            val br = BufferedReader(InputStreamReader(labelsInput))
            var line: String?
            while (true) {
                line = br.readLine()
                if (line == null) { break } else { d.labels.add(line) }
            }
            Timber.i("First class: " + d.labels.firstElement())
            br.close()
            try {
                val compatList = CompatibilityList()
                val options = Interpreter.Options().apply{
                    if(compatList.isDelegateSupportedOnThisDevice) {
                        // if the device has a supported GPU, add the GPU delegate
                        val delegateOptions = compatList.bestOptionsForThisDevice
                        //delegateOptions.inferencePreference = INFERENCE_PREFERENCE_SUSTAINED_SPEED
                        delegateOptions.inferencePreference = INFERENCE_PREFERENCE_FAST_SINGLE_ANSWER
                        delegateOptions.setQuantizedModelsAllowed(false)
                        delegateOptions.isPrecisionLossAllowed = true
                        this.addDelegate(GpuDelegate(delegateOptions))
                    } else {
                        // if the GPU is not supported, run on 4 threads
                        Timber.i("## WARNING USING CPU ##")
                        this.useXNNPACK = true
                        this.numThreads = 4
                    }
                }
                d.tfliteModel = loadModelFile(assetManager, modelFilename)
                d.tfLite = Interpreter(d.tfliteModel!!, options)
            } catch (e: Exception) {
                throw RuntimeException(e)
            }
            d.confThreshold = confThreshold
            d.isModelQuantized = isQuantized
            d.scalingFactor = inputSize.toFloat()
            d.modelType = modelType

            // Pre-allocate buffers.
            val numBytesPerChannel = if (isQuantized) 1 else 4
            d.inputSize = inputSize
            d.imgData =
                ByteBuffer.allocateDirect(1 * d.inputSize * d.inputSize * 3 * numBytesPerChannel)
            d.imgFloatArray = FloatArray(d.inputSize * d.inputSize * 3)
            d.imgData.order(ByteOrder.nativeOrder())
            d.intValues = IntArray(d.inputSize * d.inputSize)

            // yolov8 case (20² + 40² + 80²) = 8400
            d.outputBox = ((inputSize / 32).toDouble().pow(2.0) + (inputSize / 16).toDouble()
                .pow(2.0) + (inputSize / 8).toDouble().pow(2.0)).toInt()


            if (d.isModelQuantized) {
                val inpten = d.tfLite!!.getInputTensor(0)
                d.inp_scale = inpten.quantizationParams().scale
                d.inp_zero_point = inpten.quantizationParams().zeroPoint
                val oupten = d.tfLite!!.getOutputTensor(0)
                d.oup_scale = oupten.quantizationParams().scale
                d.oup_zero_point = oupten.quantizationParams().zeroPoint
            }
            val shape = d.tfLite!!.getOutputTensor(0).shape()

            val numClass: Int
            when (modelType) {
                "detection" -> {
                    numClass = shape[shape.size - 2] - 4
                    d.outData =
                        ByteBuffer.allocateDirect(d.outputBox * (numClass + 4) * numBytesPerChannel)
                    d.preds = Array(d.outputBox) { FloatArray(shape[shape.size - 2]) }
                }
                "segmentation" -> {
                    // segmentation case (1, num_class+ num_masks+4, num_anchors)
                    val shape2 = d.tfLite!!.getOutputTensor(1).shape()
                    d.resolutionMaskW = shape2[1]
                    d.resolutionMaskH = shape2[2]
                    d.numMasks = shape2[3]
                    numClass = shape[1] - 4 - d.numMasks
                    d.outData =
                        ByteBuffer.allocateDirect(d.outputBox * (4 + numClass + d.numMasks) * numBytesPerChannel)
                    // second output: masks (1, H, W, num_masks)
                    d.outData2 =
                        ByteBuffer.allocateDirect((d.resolutionMaskW * d.resolutionMaskH * d.numMasks)* numBytesPerChannel)
                    // d.masks = Mat(d.numMasks, d.resolutionMaskW * d.resolutionMaskH, CvType.CV_32F)
                    d.outData2.order(ByteOrder.nativeOrder())
                    d.preds = Array(d.outputBox) { FloatArray(shape[1]) }
                }
                else -> {
                    throw RuntimeException("model type $modelType not recognized, possible types: [detection, segmentation]")
                }
            }
            d.numClass = numClass
            d.outData.order(ByteOrder.nativeOrder())

            return d
        }
        private const val MASK_SCALE_FACTOR = 4
        private const val DEBUG_PROFILING = true
    }
}